### ISSUE TITLE
Getting error when creating a new cluster

### ISSUE THREAD

**Issue by FuzzOli87** [lae/ansible-role-proxmox#91]
Hello!

Thanks for all your hard work on this. I've been setting up my cluster and it runs fine all the way until it has to add a new node to a cluster. I see that the cluster is created and the playbook passes all the tests and reports that things were changed. However, the cluster never gets any new nodes and it actually throws and error.

I get this error:
`TASK ERROR: format error at /usr/share/perl5/PVE/JSONSchema.pm line 1866 type: Invalid schema definition. optional: Invalid schema definition. description: Invalid schema definition. format: property is missing and it is not optional`

Let me know what logs or whatever you need to debug.

**Comment by lae** [comment]
What task is that error occurring on?

**Comment by FuzzOli87** [comment]
@lae It's occurring on `clusterjoin`

**Comment by lae** [comment]
Can you please post your group/role variables?

**Comment by FuzzOli87** [comment]
@lae Sure, it's pretty plain, mostly defaults:

```
pve_group: pve_familia
pve_reboot_on_kernel_update: true
pve_run_system_upgrades: true
pve_cluster_enabled: yes
pve_cluster_clustername: familia
```

all.yaml
```
ntp_manage_config: true
ntp_servers:
  - time-a-g.nist.gov iburst
  - time-a-wwv.nist.gov iburst
  - time-a-b.nist.gov iburst
  - utcnist.colorado.edu iburst
  - time-b-wwv.nist.gov	iburst
```

**Comment by lae** [comment]
Thanks. When I get a chance (may not be super soon) I'll try reproducing it locally.

Also just to be clear (because there is no `clusterjoin` task), do you mean the `Add node to Proxmox cluster` task (that is after `Identify what host we're working with (inside outer loop)`)?

**Comment by FuzzOli87** [comment]
> Thanks. When I get a chance (may not be super soon) I'll try reproducing it locally.
> 
> Also just to be clear (because there is no `clusterjoin` task), do you mean the `Add node to Proxmox cluster` task (that is after `Identify what host we're working with (inside outer loop)`)?

Yes. I'm sorry I thought you meant the task that shows up on the Proxmox GUI. Ansible actually doesn't error out. It goes through as if the nodes were added to the cluster. But when I log into the GUI, they were not added and that error is what displays for the "clusterjoin" proxmox task at the bottom.

**Comment by Moritz0708** [comment]
I have the same Error. The 2 Servers not Join the Proxmox Cluster.


Error:
`changed: [Proxible-2] => {
    "changed": true,
    "cmd": [
        "pvecm",
        "add",
        "123.456.789.012",
        "-use_ssh",
        "-link0",
        "123.456.789.013"
    ],
    "delta": "0:00:01.808788",
    "end": "2020-05-15 08:51:08.420647",
    "invocation": {
        "module_args": {
            "_raw_params": "pvecm add 123.456.789.012 -use_ssh -link0 123.456.789.013 ",
            "_uses_shell": false,
            "argv": null,
            "chdir": null,
            "creates": "/etc/pve/corosync.conf",
            "executable": null,
            "removes": null,
            "stdin": null,
            "stdin_add_newline": true,
            "strip_empty_ends": true,
            "warn": true
        }
    },
    "rc": 0,
    "start": "2020-05-15 08:51:06.611859",
    "stderr": "format error at /usr/share/perl5/PVE/JSONSchema.pm line 1866 format: property is missing and it is not optional description: Invalid schema definition. optional: Invalid schema definition. type: Invalid schema definition.",
    "stderr_lines": [
        "format error at /usr/share/perl5/PVE/JSONSchema.pm line 1866 format: property is missing and it is not optional description: Invalid schema definition. optional: Invalid schema definition. type: Invalid schema definition."
    ],
    "stdout": "",
    "stdout_lines": []
}
`

**Comment by FuzzOli87** [comment]
@Moritz0708 How did you get this output? I tried messing around with the debugger but couldn't get the command variables to show up.

**Comment by mirasystemsmd** [comment]
Enter -vvv behind the Ansible Playbook command and you will see everything :) @FuzzOli87

**Comment by Moritz0708** [comment]
Enter -vvv behind the Ansible Playbook command and you will see everything :) @FuzzOli87

**Comment by FuzzOli87** [comment]
@lae I did some digging through the source code as well as tried some different configurations for the command. It feels like this isn't a bug in the command itself.
```
root@mymachine:~# pvecm add 192.168.0.xxx -use_ssh=yes -link0 192.168.0.xxx
format error at /usr/share/perl5/PVE/JSONSchema.pm line 1866 format: property is missing and it is not optional type: Invalid schema definition. description: Invalid schema definition. optional: Invalid schema definition.```

Throws out an error, even running the command directly on the host.

If I run the same command with ```-use_ssh=no``` or without ```-use_ssh``` at all, it works.

In the code, there is an entire section dedicated to when the ```use_ssh``` flag is false.
```
	    if (!$param->{use_ssh}) {
		my $password = PVE::PTY::read_password("Please enter superuser (root) password for '$host': ");

		delete $param->{use_ssh};
		$param->{password} = $password;

		my $local_cluster_lock = "/var/lock/pvecm.lock";
		PVE::Tools::lock_file($local_cluster_lock, 10, \&PVE::Cluster::Setup::join, $param);

		if (my $err = $@) {
		    if (ref($err) eq 'PVE::APIClient::Exception' && defined($err->{code}) && $err->{code} == 501) {
			$err = "Remote side is not able to use API for Cluster join!\n" .
			       "Pass the 'use_ssh' switch or update the remote side.\n";
		    }
		    die $err;
		}
		return; # all OK, the API join endpoint successfully set us up
	    }
```

So the error has to happen when it's set. I followed the code and my hunch is that it happens in this line:
![image](https://user-images.githubusercontent.com/12246915/82097476-94112a80-96c8-11ea-8d99-5c7af3246e6f.png)

The ```link0``` flag doesn't get used at all when use_ssh is not set. I just traced the code as much as I could without running it. 

Long story short, I think it's a bug! 

Thanks @Moritz0708 for letting me know how to use that. I'm just now picking up Ansible. That helped me getting the params for the command.

**Comment by FuzzOli87** [comment]
I created a post in the forums asking about the command failure: https://forum.proxmox.com/threads/adding-node-via-pvecm-fails-due-to-schema-failure.69882/

It doesn't seem to be anything we are doing wrong with the command itself. 🤷  We will find out!

**Comment by FuzzOli87** [comment]
https://pve.proxmox.com/pipermail/pve-devel/2020-May/043522.html

I was close! A bug was already filed for it. If y'all check out that thread, they also explain that the option is going away.

**Comment by lae** [comment]
Thanks a lot for looking into this, Daniel.

Did the thread get deleted? It's not showing up as a valid thread anymore (I saw it when you posted it but didn't see the replies) and search isn't turning anything up either.

Based on that latest patch however, I am assuming that this is fixed upstream with a package update. If that's correct, we can close this.

**Comment by FuzzOli87** [comment]
@lae No, it's still there. I believe you have to register for the forums to see the post.

Closing since this is a known bug.

**Comment by Moritz0708** [comment]
The Forum Theard: https://forum.proxmox.com/threads/adding-node-via-pvecm-fails-due-to-schema-failure.69882/ is not exists. I have registred.

**Comment by FuzzOli87** [comment]
@Moritz0708 @lae My mistake, looks like it's waiting approval to be displayed publicly.
![image](https://user-images.githubusercontent.com/12246915/82129785-ef0b5600-978a-11ea-825c-8c5a64b5d13d.png)

**Comment by Moritz0708** [comment]
Now its Public @lae @FuzzOli87, but where Fix the Problem @lae or Proxmox?

**Comment by lae** [comment]
the fix to keep the role working is upstream, which should hit PVE's packages soon (maybe it already has? I haven't checked), but since Thomas says it can be done via the API now I'll have to look into maybe making an ansible module under `library/` for cluster management actions to replace the `pvecm` command (and possibly others) in `tasks/add_node.yml`. But that'll be a separate feature request.

**Comment by FuzzOli87** [comment]
@lae Would that entail just running the command without the use_ssh flag? Looks like the Proxmox code uses the API if you don't use the flag.

Is there documentation for the Proxmox API? 

Locally @Moritz0708 , I just removed the use_ssh flag from the task. Going to try that today.

**Comment by lae** [comment]
Uh, I'm not sure it's that simple. The role sets up SSH keys etc in the first place amongst all the hosts specifically so that it can run the cluster joins, there might be some other auth procedure when performing it over the API (I haven't checked and likely won't for a while, but API docs are at https://pve.proxmox.com/pve-docs/api-viewer/index.html) that the role may have to configure. And then it sounds like we'd have to remove the ssh cluster configuration steps as well, since they would be unnecessary (unless SSH connectivity is still used for some other purpose, but I'd have to double check that too).

**Comment by FuzzOli87** [comment]
@lae Ah, ok. I'll have to dig in when I have time. Thanks for linking the API docs.

**Comment by Moritz0708** [comment]
@lae @FuzzOli87  If this command is executed on one joining Node manually, it works.

pvecm add 123.456.678.901 -use_ssh


Result: If Link0 is omitted, it works

**Comment by lae** [comment]
You're still using SSH with that command, which, if we were to use the API for cluster management, we wouldn't be doing (as far as I understand).

The `-link0` parameter specifies the cluster network. Sure, the command without `-link0` may work for you if your cluster network is the same as your management network (`123.456.678.901` in your example), but for many production environments that is not the case, so leaving out the parameter is not an option for this role. But this particular regression, as stated earlier, is fixed upstream and we just have to wait until the relevant package is updated.

**Comment by lae** [comment]
Travis CI failure demonstrating the issue: https://travis-ci.org/github/lae/ansible-role-proxmox/jobs/688608251#L1480 (CI failures from earlier were because of an unrelated test suite issue that took me a bit to get around to fixing, but normally issues like this cluster joining issue would be caught by it)

Doesn't seem like the fix has made it into the PVE repos yet. Gonna wait a few more days for it. Leaving this open for now in case other folks run into this bug.

**Comment by lae** [comment]
@FuzzOli87 @Moritz0708 Since it seems like upstream is taking a while to push an update, I went ahead and cut v1.6.4 to apply the patch ourselves. Please update if you're still having issues.
